raiden/api/python.py

```diff
@@ -310,27 +337,49 @@ def channel_batch_close(
             partner_addresses,
         )
 
-        for channel_state in channels_to_close:
-            channel_close = ActionChannelClose(
-                registry_address,
-                token_address,
-                channel_state.identifier,
-            )
-            self.raiden.handle_state_change(channel_close)
-
-        msg = 'After {} seconds the deposit was not properly processed.'.format(
-            poll_timeout
-        )
-        channel_ids = [channel_state.identifier for channel_state in channels_to_close]
-
-        with gevent.Timeout(poll_timeout, EthNodeCommunicationError(msg)):
-            waiting.wait_for_close(
-                self.raiden,
-                registry_address,
-                token_address,
-                channel_ids,
-                self.raiden.alarm.wait_time,
-            )
+        # If concurrent operations are happening on one of the channels, fail entire
+        # request. TODO: we could also prevent failing entire request and continue
+        # with those channels that are available, callers of the api will then have
+        #  to deal with partial results.
+        with ExitStack() as stack:
+            # Put all the locks in this outer context so that the netting channel functions
+            # don't release the locks when their context goes out of scope
+            for channel_state in channels_to_close:
```

Comment:
The side effect must not happen before the the state change `ActionChannelClose` is handled, this would break Raiden's recovery guarantees. Dispatching the state change saves it to the WAL, allowing the action to be re-executed if needed on restart.

The correct approach is to dispatch the action, and let the [event handler call close](https://github.com/raiden-network/raiden/blob/1f17a15c518a4af65f03c015cf13c63ef755150d/raiden/raiden_event_handler.py#L150-L178), the lock should be held individually for each channel.

Reply:
the re-entrant locks in each channel will protect those critical areas. These locks for now serve the additional purpose of protecting the API from repetitive calls (wastes gas) until we get confirmation of success from the network.

Reply:
The problem here is not the concurrent transactions, but resilience to faults (e.g. power outages, OOM kill).

If a user calls `channel_batch_close` with a list of two channels, and the process is killed after the first channel is closed but prior to sending the transaction to the second, we end up in a bad situation. There will be no information in the node's log explaining "why" the channel was closed, and the second channel will not be closed on restart.

To fix this we must go through the write-ahead-log machinery and close the channels on the event handler. The WAL guarantees that either we start the operation and retry on restart or don't start it at all.

Note: `ActionChannelClose` doesn't working with batches of channels, so it's not perfect, but making it work in batches is another problem.

Reply:
@carllin correctly pointed out the code is correct. The loop is only holding the locks for all channels before dispatching the state changes, if all locks are successfully held then the WAL is used.